pkg/util/helper/lua.go

```diff
@@ -0,0 +1,431 @@
+package helper
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"time"
+
+	lua "github.com/yuin/gopher-lua"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	luajson "layeh.com/gopher-json"
+
+	configv1alpha1 "github.com/karmada-io/karmada/pkg/apis/config/v1alpha1"
+	workv1alpha2 "github.com/karmada-io/karmada/pkg/apis/work/v1alpha2"
+)
+
+// VM Defines a struct that implements the luaVM
+type VM struct {
+	// UseOpenLibs flag to enable open libraries. Libraries are disabled by default while running, but enabled during testing to allow the use of print statements
+	UseOpenLibs bool
+}
+
+// GetReplicas returns the desired replicas of the object as well as the requirements of each replica by lua script
+func (vm VM) GetReplicas(obj *unstructured.Unstructured, script string) (replica int32, requires *workv1alpha2.ReplicaRequirements, err error) {
+	l := lua.NewState(lua.Options{
+		SkipOpenLibs: !vm.UseOpenLibs,
+	})
+	defer l.Close()
+	// Opens table library to allow access to functions to manipulate tables
+	err = vm.setLib(l)
+	if err != nil {
+		return 0, nil, err
+	}
+	// preload our 'safe' version of the OS library. Allows the 'local os = require("os")' to work
+	l.PreloadModule(lua.OsLibName, SafeOsLoader)
+
+	ctx, cancel := context.WithTimeout(context.Background(), 100*time.Second)
```

Comment:
Is this the lua script execution timeout interval? Would this default be too large, I saw the default value is one second in argo-cd project.